src/forward/array.jl

```diff
@@ -0,0 +1,25 @@
+using LinearAlgebra
+
+zerolike(x::AbstractArray) = zerolike.(x)
+
+@tangent function (A::Type{<:Array})(::UndefInitializer, sz::Integer...)
+  x = A(UndefInitializer(), sz...)
+  x, (_...) -> zerolike(x)
+end
+
+@tangent length(A::AbstractArray) = length(A), _ -> 0
+@tangent size(A::AbstractArray, i::Integer) = size(A, i), (_, _) -> 0
+@tangent size(A::AbstractArray) = size(A), _ -> zerolike(size(A))
+
+@tangent Base.vect(xs...) = Base.vect(xs...), Base.vect
+
+@tangent first(x) = first(x), ẋ -> first(ẋ)
+
+@tangent setindex!(x::AbstractArray, v, inds...) =
+  setindex!(x, v, inds...), (ẋ, v̇, _...) -> setindex!(ẋ, v̇, inds...)
+
+@tangent mul!(C, A, B) = mul!(C, A, B), (Ċ, Ȧ, Ḃ) -> Ċ .+= Ȧ*B .+ A*Ḃ
```

Comment:
My instinct says that adding to it in place is correct. Because of accumulatimg from multiple places.

But I would need to sit down and think to find an example

Reply:
I can't think of one. Here's a counter-example, where finite differences agrees with the overwriting version:

```julia
function foo(x)
    c = 2 * x
    b = x * c
    mul!(c, b, x)
    d = c * x
    return d
end

function foo_with_pushforward(x, Δx)
    c = 2 * x
    ∂c = 2 * Δx
    b = x * c
    ∂b = Δx * c + x * ∂c
    mul!(c, b, x)
    ∂c .= ∂b * x + b * Δx # overwriting
    d = c * x
    ∂d = ∂c * x + c * Δx
    return d, ∂d
end

function foo_with_pushforward2(x, Δx)
    c = 2 * x
    ∂c = 2 * Δx
    b = x * c
    ∂b = Δx * c + x * ∂c
    mul!(c, b, x)
    ∂c .+= ∂b * x + b * Δx # in-place accumulation
    d = c * x
    ∂d = ∂c * x + c * Δx
    return d, ∂d
end

julia> using FiniteDifferences, LinearAlgebra

julia> x, Δx = Float64[1 2; 3 4], Float64[1 4; 3 2];

julia> foo_with_pushforward(x, Δx)[2] ≈ jvp(central_fdm(5, 1), foo, (x, Δx))
true

julia> foo_with_pushforward2(x, Δx)[2] ≈ jvp(central_fdm(5, 1), foo, (x, Δx))
false
```

Reply:
Yeah, I think Seth's right here. The `+` is a reflex from reverse mode.